### ISSUE TITLE
Removing the USE node after removing the DEF node causes an error.

### ISSUE THREAD

**Issue by microaaron** [x3dom/x3dom#1221]
[Example: https://jsitor.com/F4zmIMSN9y](https://jsitor.com/F4zmIMSN9y)

[Key Codes: var doc = child.findX3DDoc();](https://github.com/x3dom/x3dom/blob/master/src/X3DDocument.js#L782)

```
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <script type="text/javascript" src="https://x3dom.org/download/1.8.2/x3dom-physics.js"></script>
    <link rel='stylesheet' type='text/css' href='https://x3dom.org/release/x3dom.css' />
    <title></title>
  </head>
  <body>
    <x3d showStat="false" showLog="true" x="0px" y="0px" width="900" height="900px">
      <scene id="SCENE" DEF="SCENE">

        <Group id="Gp1" DEF="Gp1">
        </Group>

        <Group id="Gp2" USE="Gp1">
        </Group>

      </scene>
    </x3d>

    <script>

      setTimeout(() => {
        document.getElementById("Gp1").remove();
      }, 1000);
      
      setTimeout(() => {
        document.getElementById("Gp2").remove();
      }, 1500);

    </script>
  </body>
</html>

```

**Comment by microaaron** [comment]
Maybe this problem has been fixed.

**Comment by microaaron** [comment]
[#1201](https://github.com/x3dom/x3dom/pull/1201)  This fix does not seem to solve the problem.

**Comment by andreasplesch** [comment]
https://jsitor.com/jpkCDaFsQi uses the latest dev version. Do you still the problem ? There do not seem to be errors.

**Comment by microaaron** [comment]
Thanks, https://github.com/x3dom/x3dom/pull/1201 really solved the problem.

[Key Codes:  return this._nameSpace ? this._nameSpace.doc : null;](https://github.com/x3dom/x3dom/blob/master/src/nodes/Core/X3DNode.js#L318)

**Comment by microaaron** [comment]
@andreasplesch Since _nameSpace and _x3domNode are emptied, if an x3d node's DEF DOM node is removed, it's hard to change its attributes.

[https://jsitor.com/iYdiuv8LmF](https://jsitor.com/iYdiuv8LmF)

```
<!DOCTYPE html>
<html>

<head>
	<meta charset="utf-8">
	<script type="text/javascript" src="https://x3dom.netlify.app/x3dom.debug.js"></script>
	<link rel='stylesheet' type='text/css' href='https://x3dom.netlify.app/x3dom.css' />
	<title></title>
</head>

<body>
	<x3d showStat="false" showLog="true" x="0px" y="0px" width="900" height="900px">
		<scene id="SCENE" DEF="SCENE">

			<Transform id="Gp1" DEF="Gp1" translation='-1 0 0'>
				<Shape id="Shape1" DEF="Shape1">
					<Appearance DEF="Appearance">
						<Material DEF="Material" emissiveColor='0.0 0.0 1.0'></Material>
					</Appearance>
					<Box DEF="Box" size='1 1 1'></Box>
				</Shape>
			</Transform>

			<Transform translation='2 0 0'>
				<Transform id="Gp2" USE="Gp1">
				</Transform>
			</Transform>

		</scene>
	</x3d>

	<script>
      var Gp1,Gp2;      
      setTimeout(() => {
        Gp = document.getElementById("Gp1")._x3domNode;
        document.getElementById("Gp1").setFieldValue("translation", new x3dom.fields.SFVec3f(-1,1,0));
        /* USE node has no setFieldValue */
        //document.getElementById("Gp2").setFieldValue("translation", new x3dom.fields.SFVec3f(-1,-2,0));
      }, 1000);
      
      setTimeout(() => {
        Gp._vf.translation.y=-1;
        Gp.fieldChanged("translation");
        Gp._nameSpace.doc.needRender = true;
      }, 1500);  
      
      setTimeout(() => {
        document.getElementById("Gp1").remove();
      }, 2000);  
      
      setTimeout(() => {
        Gp._vf.translation.y=2;
        Gp.fieldChanged("translation");
        /* _nameSpace = null  Move the mouse to update the screen */
        //Gp._nameSpace.doc.needRender = true; 
      }, 2500);  

	</script>
</body>

</html>
```

**Comment by andreasplesch** [comment]
I understand the problem. It should be possible to access the fields of a USE node even after the DEF node is removed, using setFieldValue (attributes are not available anyways). What would be the practical use case ?

https://github.com/x3dom/x3dom/blob/master/src/X3DDocument.js#L700 in .removeX3DOMBackendGraph seems to be the place which would need attention. I think the DEF node can only be removed from the namespace if there are no USE nodes. And the node should be removed from the namespace even if it a USE node if it is the only instance. In other words, a node should only be removed from the namespace if it has one (or zero?) parents.

But .setFieldValue would need to work for USE nodes first.

**Comment by microaaron** [comment]
@andreasplesch Combined with #1220, here is my suggested modification.
Move the codes of removing nameSpace into parentRemoved()

https://github.com/microaaron/x3dom/blob/%231220%2C%231221/src/X3DDocument.js#L700
```
        //do not remove node from namespace if it was only "USE"d
        //if ( nameSpace && !( domNode.getAttribute( "use" ) || domNode.getAttribute( "USE" ) ) )
        if ( domNode.getAttribute( "use" ) || domNode.getAttribute( "USE" ) )
        {
            delete domNode._x3domNode;
        }
```

https://github.com/microaaron/x3dom/blob/%231220%2C%231221/src/nodes/Core/X3DNode.js#L216
```
            parentRemoved : function ( parent )
            {
                // attention: overwritten by concrete classes
                if ( this._parentNodes.length == 0 )
                {
                    //x3dom.debug.logInfo( this.typeName() + ": " + this._DEF + " is no longer used." );
                    for ( var i = this._childNodes.length - 1;i >= 0;i-- )
                    {
                        this.removeChild( this._childNodes[ i ] ); //remove this node from child's _parentNodes then trigger child.parentRemoved()
                    }
                    var nameSpace = this._nameSpace;
                    if ( nameSpace )
                    {
                        nameSpace.removeNode( this._DEF );
                        //remove imported node from namespace
                        var superInlineNode = nameSpace.superInlineNode;
                        if ( superInlineNode && superInlineNode._nameSpace )
                        {
                            var imports = superInlineNode._nameSpace.imports;
                            var exports = nameSpace.exports;
                            var inlineDEFMap = imports.get( superInlineNode._DEF );
                            if ( inlineDEFMap )
                            {
                                exports.forEach( function ( localDEF, exportedAS )
                                {
                                    if ( this._DEF == localDEF )
                                    {
                                        inlineDEFMap.forEach( function ( importedDEF, importedAS )
                                        {
                                            if ( exportedAS == importedDEF )
                                            {
                                                delete superInlineNode._nameSpace.defMap[ importedAS ];
                                            }
                                        } );
                                    }
                                } );
                            }
                        }
                    }
                    if ( this._xmlNode )
                    {
                        delete this._xmlNode._x3domNode;
                        this._xmlNode = null;
                    }
                }
            },
```

**Comment by microaaron** [comment]
> But .setFieldValue would need to work for USE nodes first. The functions probably could just be attached here:
> 
> https://github.com/x3dom/x3dom/blob/79262f7cba96bbaf3e38a9727441c185510069b2/src/NodeNameSpace.js#L402

https://github.com/microaaron/x3dom/blob/%231220%2C%231221/src/NodeNameSpace.js#L400
```
            if ( n )
            {
                domNode._x3domNode = n;
                // attach X3DOM's custom field interface functions
                domNode.requestFieldRef = x3dom.requestFieldRef;
                domNode.releaseFieldRef = x3dom.releaseFieldRef;
                domNode.getFieldValue = x3dom.getFieldValue;
                domNode.setFieldValue = x3dom.setFieldValue;
            }
```

**Comment by microaaron** [comment]
@andreasplesch
https://github.com/microaaron/-1220-1221dist
It seems to be working fine. But there are still many classes that overload parentRemoved() that need to be modified。
[https://jsitor.com/4qLoOgs_VW](https://jsitor.com/4qLoOgs_VW)

**Comment by andreasplesch** [comment]
Perhaps there is a way to avoid using parentRemoved ?

**Comment by microaaron** [comment]
> Perhaps there is a way to avoid using parentRemoved ?

I suggest to modify these parentRemoved()s, by the way to fix some potential problems.

Inline.js:
https://github.com/x3dom/x3dom/blob/47610df99d1e2d46649c0a5e64b5aab6a21066a7/src/nodes/Networking/Inline.js#L159
What does global[ "_remover" ] do? It doesn't seem to do anything.

https://github.com/x3dom/x3dom/blob/47610df99d1e2d46649c0a5e64b5aab6a21066a7/src/nodes/Networking/Inline.js#L154-L161
Array members should be removed from back to front. This for loop doesn't seem to work properly, but luckily Inline seems to have only one child.

**Comment by andreasplesch** [comment]
delete global["_remover"] is (was?) apparently a trick to free memory:

https://github.com/x3dom/x3dom/blame/0d55bfd719a44846dcc0e8ae50c413a572927635/src/nodes/Networking/Inline.js#L239

Adding parentRemoved() here was part of fixing a memory leak when removing inlines:
https://github.com/x3dom/x3dom/commit/db9b51b7c326900ed78eb7957e2d9f563e498080

Right, the loop should go from n to 0.
Here as well:
https://github.com/x3dom/x3dom/blame/47610df99d1e2d46649c0a5e64b5aab6a21066a7/src/nodes/Networking/Inline.js#L111

**Comment by microaaron** [comment]
> delete global["_remover"] is (was?) apparently a trick to free memory:

I haven't heard of this trick. removeChild() returns a Boolean value. So global["_remover"] is just a Boolean value?

**Comment by andreasplesch** [comment]
The believe appears to be that simply the act of deleting a property (with any value) from the global object frees memory, probably meaning that it triggers garbage collection. Not sure where that came from.

**Comment by microaaron** [comment]
Is it necessary to trigger garbage collection immediately?

Another problem: This if() never seems to be true. Will a removed parent appear in _parentNodes?
https://github.com/x3dom/x3dom/blob/8cf82348f4fa1261b1a4d805bc357ca8bc6155d5/src/nodes/Lighting/X3DLightNode.js#L184-L186

**Comment by microaaron** [comment]
Only remove _childNodes[ 0 ]'s _nameSpace? Maybe it has only one child.
https://github.com/x3dom/x3dom/blob/23bb1572587a1a9117280d1da26ffc33bc2f1f2e/src/nodes/Networking/Inline.js#L154-L155

**Comment by andreasplesch** [comment]
> Is it necessary to trigger garbage collection immediately?

I do not know what the original motivation was. It does not seem to hurt.
 
> Another problem: This if() never seems to be true. Will a removed parent appear in _parentNodes?
> 
https://github.com/x3dom/x3dom/blob/8cf82348f4fa1261b1a4d805bc357ca8bc6155d5/src/nodes/Lighting/X3DLightNode.js#L184-L186

As this clause seems very deliberate, apparently there is (was?) a situation when parentRemoved was called before _parentNodes[] was updated.

**Comment by andreasplesch** [comment]
The Scene of the Inline is the child of Inline. Since in x3dom Inline is derived from X3DGroupingNode, in theory user code could add additional children. But in X3D Inline is not a Grouping node, so this should not happen.

**Comment by microaaron** [comment]
> As this clause seems very deliberate, apparently there is (was?) a situation when parentRemoved was called before _parentNodes[] was updated. For example, I think it could happen when a light node is the child of a transform within a transform. Then parentRemoved() is called from this parentRemoved:

Currently, after a node is removed, all its children, grandchildren, great-grandchildren. . . will execute parentRemoved, but these children's parents are not actually removed from their _parentNodes. 
So if its this._parentNodes.length === 0, should this if block be executed?
https://github.com/x3dom/x3dom/blob/8cf82348f4fa1261b1a4d805bc357ca8bc6155d5/src/nodes/Lighting/X3DLightNode.js#L184-L201

**Comment by andreasplesch** [comment]
> > As this clause seems very deliberate, apparently there is (was?) a situation when parentRemoved was called before _parentNodes[] was updated. For example, I think it could happen when a light node is the child of a transform within a transform. Then parentRemoved() is called from this parentRemoved:
> 
> Currently, after a node is removed, all its children, grandchildren, great-grandchildren. . . will execute parentRemoved, but these children's parents are not actually removed from their _parentNodes. This situation needs to be changed.

It looks like it was designed like this, probably for a reason.

> And if its this._parentNodes.length === 0, should this if block be executed?
> 
> https://github.com/x3dom/x3dom/blob/8cf82348f4fa1261b1a4d805bc357ca8bc6155d5/src/nodes/Lighting/X3DLightNode.js#L184-L201

Well, above has `.length === 1`

> 
> Similar codes exists in this file:
> https://github.com/x3dom/x3dom/blob/47610df99d1e2d46649c0a5e64b5aab6a21066a7/src/nodes/Followers/X3DFollowerNode.js#L63-L80

Follower nodes (like Interpolators) are special child nodes, which are not affected by their parents. Also, there is no (good) reason to DEF/USE them. Not sure, though, why this is only executed when there are no parents, eg. it is a root node. I suspect there is a reason. I guess, the nodebag will be updated elsewhere if the follower is a child. Then the clause here would prevent unnecessary double updating.

Apparently, there is a case when this parentRemoved() is called when this is a root node.

**Comment by microaaron** [comment]
> Well, above has `.length === 1`. Perhaps the `=== 0` case is handled somewhere else ?
cleanNodeBag() is executed whenever the DOM is removed.
https://github.com/x3dom/x3dom/blob/47610df99d1e2d46649c0a5e64b5aab6a21066a7/src/X3DDocument.js#L645-L648

**Comment by andreasplesch** [comment]
Right. And removing the DOM node is the expected way to remove such a root node.

As far as I remember, missing support for USE Light nodes has to do with how lights are collected for webgl. Probably this could be fixed but there has not been any demand probably because too many lights limit performance. I am not sure if any of the web3d examples have USE lights.

It looks like the for loops in parentRemoved could be replaced with a call to doc.cleanNodeBad() for cleaner code.

**Comment by microaaron** [comment]
Since a light cannot be USEd, is the cleanNodeBad in parentRemoved redundant? removeX3DOMBackendGraph will always be executed.

**Comment by microaaron** [comment]
> > Currently, after a node is removed, all its children, grandchildren, great-grandchildren. . . will execute parentRemoved, but these children's parents are not actually removed from their _parentNodes. This situation needs to be changed.
> 
> It looks like it was designed like this, probably for a reason.

At first, it may be taken for granted that as long as a node is removed, all its children, grandchildren, great-grandchildren... will be discarded, but even so, these children are still related to each other, which may cause the memory to not be freed.

**Comment by microaaron** [comment]
Do some tests first.
https://github.com/microaaron/-1220-1221dist
https://github.com/x3dom/x3dom/commit/067c41057c58e5c051a2f8ded389d9edd737b771

**Comment by microaaron** [comment]
@andreasplesch 
Have you tried removing a movietexture dynamically? It seems to cause errors.
Try to remove the movietexture from the example.
https://doc.x3dom.org/tutorials/basics/imagesMovies/example.html

**Comment by microaaron** [comment]
@andreasplesch Thanks for fixing the movietexture. There are still some problems. cleanNodeBag() destroys the properties of x3domNode when deleting a DEF node.
https://jsitor.com/6_F51D3Pf
I don't think removeX3DOMBackendGraph() is a good function. X3DDocument.js should not interfere with the internal properties of x3domNode.

**Comment by andreasplesch** [comment]
Yes, I think the design did not anticipate the case where a DEF node is removed before a USE node. Other than X3D actually allowing this, it does seem like a reasonable assumption in a DOM based system. After all, one you remove the DEF node, you cannot access the attributes of the node anymore.

What would be a good use case for removing a DEF node but not the USE node ? One can make the DEF node invisible instead of removing it ?

**Comment by brutzman** [comment]
There is no good use case for removing a DEF node but not deleting/handling any corresponding USE node(s).  The model becomes invalid.

There is probably a use case (so to speak) for a boolean method checking whether a DEF node has USE nodes.

Perhaps desirable to have a utility method for replacing the first USE node with the soon-to-be-deleted DEF node.

**Comment by andreasplesch** [comment]
> There is no good use case for removing a DEF node but not deleting/handling any corresponding USE node(s). The model becomes invalid.

This was the initial assumption, and I think the one followed by most X3D authors. Subsequently, there have been suggestions that DEF and USE are semantics only at initial model loading and then placements of DEF/USE nodes in the graph become indistinguishable during model execution.

> There is probably a use case (so to speak) for a boolean method checking whether a DEF node has USE nodes.

In x3dom, each node has a parentNodes property. If there is more than one parent, there is a USE node. Perhaps this could be formalized in the SAI; node.getParents() returns MFNode[], or so.

> Perhaps desirable to have a utility method for replacing the first USE node with the soon-to-be-deleted DEF node.

Not sure. [ It would mean moving the complete DEF node with all its potential children around, in the DOM, with some kind of handover. ]
